### ISSUE TITLE
Dataset-level dependencies (Lazy dependencies)

### ISSUE THREAD

**Issue by ciyer** [SwissDataScienceCenter/renku-python#706]
The main question behind this issue is how are dependencies to data expressed. This question is very general and the solution impacts many areas: how the lineage can be displayed, what the most natural way for structuring repositories, and how to write code that operates on data. Any solution needs to consider its impact on all these areas.

**Is your feature request related to a problem? Please describe.**
 How should data dependencies be described in datasets that are appended to? 

I have a project where I am analyzing flight data. Data is made available on a monthly basis. I store this raw data as datasets in repositories by year (e.g., data from 2017 in the `flights-raw-data-2017` with has one dataset, `flights-raw-2017`, with 12 files (`2017-01-flights.zip`, etc)). The same pattern holds for 2018, etc.

For actually working with the data, I apply some pre-processing to discard columns that I do not use and to convert some values into a different format. This is all stored in one repository, the `flights-preprocessed` repository. The script in this repository searches all the referenced raw-data repositories and outputs a pre-processed data file for each raw data that has not yet been processed.

I perform several kinds of analysis on the preprocessed data.

![image](https://user-images.githubusercontent.com/1196411/65500306-6704d300-debf-11e9-93bb-f34ec0306c80.png)

Currently lineage always references concrete files. This means that the arrival of new data requires manually re-running all steps of the pipeline (_pushing_ changes through the pipeline) because new data is simply unknown downstream.

Conceptually, I feel the most natural way to describe the solution is that the preprocessing script depends on the raw-data datasets (not the files). This makes it possible to detect that the addition of a new data to one of the datasets should trigger rerunning the pipeline (_pulling_ changes through the pipeline).

**Describe the solution you'd like**
I'm not exactly sure what the best solution is, but I would like to open up a discussion.

**Comment by rokroskar** [comment]
This is a great question/request - I think we need to have some more tooling around datasets in general. It seems to me that there are two general issues to solve:

1) how to tell the process running under `renku run` about the files/data in the dataset
2) how to use the dataset information when processing lineage

The first could have a very simple first solution: write a temporary file somewhere with the list of files in the dataset. We could put the location of this file into an environment variable that is available to the process. 

The second is a bit trickier. For example, I might have a dataset of "living" data files that get updated via a pipeline. I can reference the dataset with a flag like `--input-dataset` but unless I also link the run dependencies to the individual files I won't see that downstream results need to be updated. Unless the dataset is invalidated if the files change and always refers to files at a specific commit/checksum.

**Comment by ciyer** [comment]
Yes, a dataset needs to have a version/hash. This could be computed from the hash of each file it contains.

**Comment by Panaetius** [comment]
As discussed in the "2020-02-20 Renku Design meeting - Datasets" it'd be nice to have two axis of datasets, `explicit` vs `implicit`, and `permanent` vs `ephemeral`. Writing this down so it doesn't get lost.

`explicit` = user created, with user added metadata
`implicit` = generated by renku, not necessarily visible to the user, an internal abstraction. 

And all renku run commands will act on datasets (maybe call it data collections and only use "Dataset" for explicit, user created ones), simplifying implementation on our end (no separate code paths for files or datasets.

The ephemeral/permanent dimension ties into renku run outputs being treated as a dataset, with `ephemeral` ones just containing the metadata (and the workflow having to be re-run if someone else wants the data), and `permanent` ones consisting of metadata+data/files. This way, we don't need to bundle code with datasets (as some other data repositories do it) but we can instead use `input dataset` + `renku run workflow` --> `output dataset` and treat the `output dataset` as something that can be imported into other renku projects that will re-execute the code on that end when needed. And `ephemeral` would help deal with cases where the data of a dataset can't be published but the metadata should still be in the KG.

**Comment by Panaetius** [comment]
From https://github.com/SwissDataScienceCenter/renku-python/issues/787 :

Renku users would like to use the `Dataset` abstraction in other renku commands. The most obvious example is to provide the ability for something like:

```bash
renku run --input-dataset <dataset> <command>
```
In this case, renku should make the contents of `<dataset>` available to `<command>`. 

Similarly, it would be desirable to enable this behavior:

```bash
renku run --output-dataset <dataset> <command>
```
Here, outputs would automatically be added to `<dataset>`.
